### ISSUE TITLE
[Bug] Homeassistant crashes after Runtime Error in Bambu lab extension

### ISSUE THREAD

**Issue by maunsen** [greghesp/ha-bambulab#479]
### Describe the bug

I have a P1S connected via Lan mode and everything is runnning fun until homeassistant suddenly crashes with lots of errors regarding the bambulab integration.

I run Home Assistant Version 2024.2.5 in a proxmox environment. i still can connect via console and restart homeassistant.

P1S information:
Firmware: 01.05.02.00
Hardware: AP04

I am also using an AMS. 

### To Reproduce

Install the Bambulab component and connect a Bambu Lap P1S. Wait

### Expected Behaviour

Homeassistant does not crash

### What device are you using?

P1S

### Diagnostic Output

```shell
2024-03-03
20:57:00.136
ERROR
(Thread-11)
[custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:02.153
ERROR
(Thread-11)
[custom_components.bambu_labl An exception occurred calling async_set_updated _data():
2024-03-03
20:57:02.154
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type:
‹class 'Runt imeError' ›
2024-03-03
20:57:02.154
ERROR
(Thread-11)
[custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:03.509
ERROR
(Thread-9
(matt_
2024-03-03
20:57:03.510
ERROR
(Thread-9
Cmgtt.
listen thread) custon components. banbu labl Exceptept ope red 11 moset updated data () :
2024-03-03
20:57:03.
511
ERROR
(Thread-9
(matt
listen_thread)) [custom_components.bambu_lab] Exception data: Event loop is closed
2024-03-03
20:57:05.172
ERROR
(Thread-11)
[custom_components.bambu_labl An exception occurred calling async_set_updated _data():
2024-03-03
20:57:05.173
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: ‹class
'Runt imeError' ›
2024-03-03
20:57:05.173
ERROR
(Thread-11)
[custom
_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:05.798
ERROR
(Thread-9 (natt listen thread) custon components. banbu lab Exceptiot te cred it neror set updated data ():
2024-03-03
20:57:05.799
ERROR

2024-03-03
20:57:05.799
ERROR
(Thread-9 (mott
_listen_thread)) [custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:07.189
ERROR
(Thread-11) [custom_components.bambu_labl An exception occurred calling async_set_updated_data():
2024-03-03
20:57:07.190
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: <class
'Runt imeError' ›
2024-03-03
20:57:07.191
ERROR
(Thread-11)
[custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:09.207
ERROR
(Thread-11)
[custom_components.bambu_labl An exception occurred calling async_set_updated _data@:
2024-03-03
20:57:09.208
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: ‹class 'RuntimeError' ›
2024-03-03
20:57:09.208
ERROR
(Thread-11)
[custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:09.837
ERROR
(Thread-9
_listen_thread)) [custom_components.bambu_lab] An exception occurred calling async_set_updated_data():
2024-03-03
20:57:09.838
ERROR
(Thread-9 (matt.
listen_thread)) [custom_components.bambu_labl Exception type: <class 'RuntimeError' ›
2024-03-03
20:57:09.839
ERRORI
(Thread-9 (matt.
listen_thread)) [custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:11.224
ERROR
(Thread-11)
[custom_components.bambu_labl An exception occurred calling async_set_updated_data():
2024-03-03
20:57:11.224
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: ‹class
'Runt imeError' >
2024-03-03
20:57:11.225
ERROR
(Thread-11)
[custom
components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:11.864
ERROR
(Thread-9
Cmgtt.
_listen_thread)) [custom_components.bambu_labl An exception occurred calling async_set_updated_data():
2024-03-03
20:57:11.865
ERROR
(Thread-9 (matt.
listen_thread)) [custom_components.bambu_labl Exception type: ‹class 'RuntimeError' ›
2024-03-03
20:57:11.865
ERROR
(Thread-9 (matt.
_listen_thread)) [custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:13.240
ERROR
(Thread-11) [custom_components.bambu_labl An exception occurred calling async_set_updated_data():
2024-03-03
20:57:13.241
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: <class
'Runt imeError' ›
2024-03-03
20:57:13.242
ERROR
(Thread-11)
[custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:13.894
ERROR
(Thread-g
(matt_listen_thread)) [custom_components.bambu_lab] An exception occurred calling async_set_updated _data():
2024-03-03
20:57:13.894
ERROR
(Thread-9
(mytt_listen_thread)) [custom_components.bambu_labl Exception type: <class 'RuntimeError' >
2024-03-03
20:57:13.895
ERROR
(Thread-9
Cmgtt.
_listen_thread)) [custom_components.bambu_labl Exception data: Event loop is closed
2024-03-03
20:57:15.259
ERROR
(Thread-11)
[custom_components.bambu_labl An exception occurred calling async_set_updated_data():
2024-03-03
20:57:15.260
ERROR
(Thread-11)
[custom_components.bambu_labl Exception type: ‹class 'Runt imeError'›
2024-03-03
20:57:15.260
ERROR
(Thread-11) [custom_components.bambu_labl Exception data: Event loop is
```

**Comment by AdrianGarside** [comment]
Can you provide logs leading up to the errors? The log snippet you provided doesn't have any context as to what might have caused the problem.

**Comment by AdrianGarside** [comment]
And with DEBUG logging enabled for the integration - so we can get more information about what the integration was doing leading up to the failure.

**Comment by maunsen** [comment]
Thank's for the response. I just activated debug logging. When it happens again, I will provide the logs.

**Comment by maunsen** [comment]
Hi, 
I just added the logs. 
At arount 1:24 this morning homeassistant crashed again. It was during a long overnight print.
It seems only to happen when I am printing. It never occured while the printer was idle.
I hope this helps.
If you need anything else, let me know.
[home-assistant.log](https://github.com/greghesp/ha-bambulab/files/14554528/home-assistant.log)

**Comment by maunsen** [comment]
Thanks, just upgraded to Version  2.0.16-dev1

But why did HA shut down? I didn't tell it to.

**Comment by maunsen** [comment]
I just had the next crash with Version 2.0.16-dev1
[20240311-ha.log](https://github.com/greghesp/ha-bambulab/files/14561088/20240311-ha.log)

**Comment by AdrianGarside** [comment]
The trigger seems to be an automation you have setup. What does the automation do? The automation is running on the main HA thread and looks to be getting a runtime error as it's querying state from the integration. I've seen this occasionally and have added protection in the integration when it's the one making calls into HA but this is being triggered outside the integrations control so I have to root cause the problem and find a way to avoid it to make this usage case safe:

RuntimeError: list changed size during iteration


...
RuntimeError: Executor shutdown has been called
2024-03-11 16:43:32.206 ERROR (MainThread) [homeassistant.components.automation.bambu_lab_p1_spaghetti_detection] Bambu Lab P1 - Spaghetti Detection: Choose at step 1: choice 3: Error executing script. Unexpected error for if at pos 2: Executor shutdown has been called
...
RuntimeError: Executor shutdown has been called
2024-03-11 16:43:32.307 ERROR (MainThread) [homeassistant.components.automation.bambu_lab_p1_spaghetti_detection] Bambu Lab P1 - Spaghetti Detection: Error executing script. Unexpected error for choose at pos 1: Executor shutdown has been called
...
RuntimeError: Executor shutdown has been called
2024-03-11 16:43:32.442 ERROR (MainThread) [homeassistant.components.automation.bambu_lab_p1_spaghetti_detection] While executing automation automation.bambu_lab_p1_spaghetti_detection
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/runner.py", line 188, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 672, in run_until_complete
    self.run_forever()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 639, in run_forever
    self._run_once()
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1961, in _run_once
    handle = heapq.heappop(self._scheduled)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: list changed size during iteration

During handling of the above exception, another exception occurred:

**Comment by AdrianGarside** [comment]
New release with a bit more logging around the HMS error list setting/retrieval to see if that proves it's involved in this:
https://github.com/greghesp/ha-bambulab/releases/tag/v2.0.16-dev2

**Comment by AdrianGarside** [comment]
Please try this fix:
https://github.com/greghesp/ha-bambulab/releases/tag/v2.0.16-dev4

**Comment by gustafssone** [comment]
I had plenty of issues in 2.0.15, but it seems to be solved with the fix v2.0.16-dev4. 
HA 2024.3 and ha-bambulab v2.0.15 had no issues for me.
HA 2024.4 and ha-bambulab v2.0.15 crashed.
HA 2024.4 and ha-bambulab v2.0.16-dev4 no issues (so far).
Great work!

Log for HA 2024.4 and ha-bambulab v2.0.15:
`
2024-04-16 08:34:08.769 ERROR (Thread-5) [custom_components.bambu_lab] An exception occurred calling async_set_updated_data():
2024-04-16 08:34:08.769 ERROR (Thread-5) [custom_components.bambu_lab] Exception type: <class 'RuntimeError'>
2024-04-16 08:34:08.769 ERROR (Thread-5) [custom_components.bambu_lab] Exception data: Non-thread-safe operation invoked on an event loop other than the current one
2024-04-16 08:34:09.523 ERROR (Thread-3 (mqtt_listen_thread)) [custom_components.bambu_lab] An exception occurred calling async_set_updated_data():
2024-04-16 08:34:09.523 ERROR (Thread-3 (mqtt_listen_thread)) [custom_components.bambu_lab] Exception type: <class 'RuntimeError'>
2024-04-16 08:34:09.523 ERROR (Thread-3 (mqtt_listen_thread)) [custom_components.bambu_lab] Exception data: Non-thread-safe operation invoked on an event loop other than the current one
2024-04-16 08:37:06.838 ERROR (Thread-5) [custom_components.bambu_lab] An exception occurred calling async_set_updated_data():
2024-04-16 08:37:06.838 ERROR (Thread-5) [custom_components.bambu_lab] Exception type: <class 'RuntimeError'>
2024-04-16 08:37:06.838 ERROR (Thread-5) [custom_components.bambu_lab] Exception data: Event loop is closed

`

**Comment by AdrianGarside** [comment]
Closing this as resolved. Please open a new issue (with debug logs) if you still hit any further instability to help me keep repros prior to this fix vs after.
